File: src/NnCase.Transforms/Transform.cs
using System;
using System.Collections.Generic;
using NnCase.IR;

namespace NnCase.Transforms
{
    public class TransformContext
    {
        public Graph Graph { get; }

        public List<Node> MatchedNodes { get; } = new List<Node>();

        public List<InputConnector> Inputs { get; } = new List<InputConnector>();

        public List<OutputConnector> Outputs { get; } = new List<OutputConnector>();

        public TransformContext(Graph graph)
        {
            Graph = graph;
        }
    }

    public abstract class Transform
    {
        protected virtual bool SkipSelfContainedCheck => false;

        public bool TryMatch(Node node, TransformContext context)
        {
            if (OnTryMatch(node, context))
            {
                if (!SkipSelfContainedCheck)
                {
                    // there exist input connectors out of the subgraph
                    foreach (var input in node.Inputs)
                    {
                        if (input.Connection != null)
                        {
                            if (!context.Inputs.Contains(input) && !context.MatchedNodes.Contains(input.Connection.Owner))
                                return false;
                        }
                    }

                    // there exist output connectors out of the subgraph
                    foreach (var output in node.Outputs)
                    {
                        foreach (var conn in output.Connections)
                        {
                            if (!context.Outputs.Contains(output) && !context.MatchedNodes.Contains(conn.Owner))
                                return false;
                        }
                    }
                }

                return true;
            }

            return false;
        }

        public abstract void Process(TransformContext context);

        protected abstract bool OnTryMatch(Node node, TransformContext context);

        public static void TransformGraph(Graph graph, IEnumerable<Transform> transforms)
        {
            bool nextPass = false;
            bool needRetry = false;
            Transform transform = null;

            var visitor = new RelayDfsVisitor(n =>
            {
                var context = new TransformContext(graph);
                if (transform.TryMatch(n, context))
                {
                    transform.Process(context);
                    needRetry = true;
                    return true;
                }

                return false;
            });

            do
            {
                nextPass = false;

                foreach (var srcTransform in transforms)
                {
                    transform = srcTransform;
                    needRetry = false;
                    visitor.Visit(graph);

                    if (needRetry)
                    {
                        nextPass = true;
                        graph.Collect();
                        break;
                    }
                }
            } while (nextPass);
        }
    }
}

File: src/transforms/llir/transform.cpp
/* Copyright 2019-2020 Canaan Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <algorithm>
#include <llir/transforms/transform.h>
#include <llir/visitor.h>

using namespace nncase;
using namespace nncase::llir;
using namespace nncase::llir::transforms;

bool transform::try_match(node &node, transform_context &context)
{
    if (on_try_match(node, context))
    {
        if (!skip_self_contained_check())
        {
            for (auto &&node : context.matched_nodes)
            {
                // there exist input connectors out of the subgraph
                for (auto &&in : node->inputs())
                {
                    if (in.connection())
                    {
                        if (std::find(std::begin(context.inputs), std::end(context.inputs), &in) == std::end(context.inputs)
                            && std::find(std::begin(context.matched_nodes), std::end(context.matched_nodes), &in.connection()->owner()) == std::end(context.matched_nodes))
                            return false;
                    }
                }

                // there exist output connectors out of the subgraph
                for (auto &&out : node->outputs())
                {
                    for (auto &&conn : out.connections())
                    {
                        if (std::find(std::begin(context.outputs), std::end(context.outputs), &out) == std::end(context.outputs)
                            && std::find(std::begin(context.matched_nodes), std::end(context.matched_nodes), &conn->owner()) == std::end(context.matched_nodes))
                            return false;
                    }
                }
            }
        }

        return true;
    }

    return false;
}

bool transform::skip_self_contained_check() const noexcept
{
    return false;
}

namespace
{
class transform_apply_visitor : public dfs_ir_visitor
{
public:
    using dfs_ir_visitor::visit;
    llir::graph *graph;
    nncase::target *target;
    bool need_retry = false;
    transforms::transform *transform;

protected:
    bool visit(node &node) override
    {
        transform_context context { *graph, *target };
        if (transform->try_match(node, context))
        {
            transform->process(context);
            need_retry = true;
            return true;
        }

        return false;
    }
};
}

std::vector<llir::input_connector *> nncase::llir::transforms::dup(xtl::span<llir::input_connector *const> connections)
{
    std::vector<llir::input_connector *> con;
    std::copy(connections.begin(), connections.end(), std::back_inserter(con));
    return con;
}

void nncase::llir::transforms::link(llir::output_connector &old_c, llir::output_connector &new_c)
{
    //new_c.attributes(old_c.attributes());
}
